### PR TITLE
Implement hash tag completion

### PR DESCRIPTION
**Pull request by orekyuu**
This PR implements completion of hashtags.
The tooted hashtag is included in the candidate.
Candidates are stored localStorage.

### CODE REVIEW

**Review by beatrix-bitrot — COMMENTED** [review]
overall this is really cool and i like it!

File: app/javascript/mastodon/components/autosuggest_textarea.js [hunk 1]
```diff
@@ -17,11 +17,11 @@ const textAtCursorMatchesToken = (str, caretPosition) => {
     word = str.slice(left, right + caretPosition);
   }
 
-  if (!word || word.trim().length < 2 || word[0] !== '@') {
+  if (!word || word.trim().length < 2 || word[0] !== prefix) {
     return [null, null];
   }
 
-  word = word.trim().toLowerCase().slice(1);
```
Comment (orekyuu) [inline app/javascript/mastodon/components/autosuggest_textarea.js]: I thought the hashtag would distinguish between uppercase and lowercase letters.
I found it to be all lowercase so fixed it.

**Comment by ykzts** [comment]
resolve conflicts.

File: app/javascript/mastodon/actions/compose.js [hunk 2]
```diff
@@ -104,6 +108,24 @@ export function submitCompose() {
...
+      }
+      const maxSize = 1000;
+      tags = tags.slice(0, maxSize);
+
+      const data = JSON.stringify(tags);
+      try {
+        localStorage.setItem('hash_tag_history', data);
+      } catch (e) {
+        //ignore
+      }
```
Comment (nolanlawson) [inline app/javascript/mastodon/actions/compose.js]: Storing large amounts of data in LocalStorage is [inefficient](https://hacks.mozilla.org/2012/03/there-is-no-simple-solution-for-local-storage/). I'd advise using IDB instead via a library like [Dexie](http://dexie.org/). You can also use it for range queries, meaning you don't need to constantly re-parse a big string of JSON and then search through it using JavaScript.

**Review by nolanlawson — CHANGES_REQUESTED** [review]

**Comment by ykzts** [comment]
@orekyuu ping

**Comment by Gargron** [comment]
Closing due to inactivity.

**Pull request closed without merging**